src/directLineStreaming.ts

```diff
@@ -0,0 +1,344 @@
+
+import {
+  Activity,
+  ConnectionStatus,
+  Conversation,
+  DirectLineOptions,
+  IBotConnection,
+  Media,
+  Message
+} from './directLine';
+
+const DIRECT_LINE_VERSION = 'DirectLine/3.0';
+const MAX_RETRY_COUNT = 3;
+const refreshTokenLifetime = 30 * 60 * 1000;
+const timeout = 20 * 1000;
+const refreshTokenInterval = refreshTokenLifetime / 2;
+
+class StreamHandler implements BFSE.RequestHandler {
+  private connectionStatus$;
+  private subscriber: Subscriber<Activity>;
+  private shouldQueue: () => boolean;
+  private activityQueue: Array<Activity> = [];
+
+  constructor(s: Subscriber<Activity>, c$: Observable<ConnectionStatus>, sq: () => boolean) {
+    this.subscriber = s;
+    this.connectionStatus$ = c$;
+    this.shouldQueue = sq;
+  }
+
+  public setSubscriber(s: Subscriber<Activity>) {
+    this.subscriber = s;
+  }
+
+  async processRequest(request: BFSE.IReceiveRequest, logger?: any): Promise<BFSE.StreamingResponse> {
+    const streams = [...request.streams];
+    const stream0 = streams.shift();
+    const activitySetJson = await stream0.readAsString();
+    const activitySet = JSON.parse(activitySetJson);
+
+    if (activitySet.activities.length !== 1) {
+      // Only one activity is expected in a set in streaming
+      this.subscriber.error(new Error('there should be exactly one activity'));
+      const r = new BFSE.StreamingResponse();
+      r.statusCode = 500;
+      return r;
+    }
+
+    const activity = activitySet.activities[0];
+
+    if (streams.length > 0) {
+      let attachments = [...activity.attachments];
+
+      let stream: BFSE.ContentStream;
+      while (stream = streams.shift()) {
+        const attachment = await stream.readAsString();
+        const dataUri = "data:text/plain;base64," + attachment;
+        attachments.push({ contentType: stream.contentType, contentUrl: dataUri });
+      }
+
+      activity.attachments = attachments;
+    }
+
+    if (this.shouldQueue()) {
+      this.activityQueue.push(activity);
+    } else {
+      this.subscriber.next(activity);
+    }
+
+    const r = new BFSE.StreamingResponse();
+    r.statusCode = 200;
+    return r;
+  }
+
+  public flush() {
+    this.connectionStatus$.subscribe(cs => { })
+    this.activityQueue.forEach((a) => this.subscriber.next(a));
+    this.activityQueue = [];
+  }
+}
+
+export class DirectLineStreaming implements IBotConnection {
+  public connectionStatus$ = new BehaviorSubject(ConnectionStatus.Uninitialized);
+  public activity$: Observable<Activity>;
+
+  private activitySubscriber: Subscriber<Activity>;
+  private theStreamHandler: StreamHandler;
+
+  private retryCount = MAX_RETRY_COUNT;
+
+  private domain = "https://directline.botframework.com/v3/directline";
+
+  private conversationId: string;
+  private token: string;
+  private streamConnection: BFSE.WebSocketClient;
+  private queueActivities: boolean;
+
+  private _botAgent = '';
+
+  constructor(options: DirectLineOptions) {
```

Comment:
This line should be:

```suggestion
  constructor(options: DirectLineStreamingOptions) {
```

Reply:
List of stuff inside:
- `token`
- `domain`
- `conversationId` (optional)
- `botAgent` (optional)